### ISSUE TITLE
DateTimeOffset.Add regression in 3.1.3

### ISSUE THREAD

**Issue by MaceWindu** [linq2db/linq2db#2512]
@viceroypenguin , looks like #2382 broke some cases that worked before.
I have following code that worked before:
```cs
db.Table.Insert(() => new Record()
{
    Field = DateTimeOffset.Now.Add(timespanValue)
});
```
now I get following error
```
'Sql.CurrentTzTimestamp.Add(...)' cannot be converted to SQL.
```

Looks like before we used client-side parameter evaluation, now it tries to convert it to server-side code and fails for some reason.

**Comment by viceroypenguin** [comment]
Yes, on my list for today. See also: #2508

**Comment by viceroypenguin** [comment]
Point of discussion: (@MaceWindu @sdanyliv)

We currently do not have support for `DateTime.Add(TimeSpan)` _or_ `DateTimeOffset.Add(TimeSpan)` in the sql, mainly because it would be difficult to translate a `TimeSpan` into a parameter to send to any sql server.  

As you noted, moving `DateTimeOffset.Now` and `Sql.CurrentTzTimestamp` to the server means that the above code breaks. This is because we're now attempting to call the `TimeSpan` version of `.Add()`. The above code would be broken from the start if it were `DateTime.Now.Add(timespanValue)`, for the same reason.

Options:
1. [Non-breaking, but less pretty] Revert moving `DateTimeOffset.Now` to the server, leaving `Sql.CurrentTzTimestamp` as the only way to get the server TZ-based timestamp. Pros: non-breaking. Cons: difference in behavior between `DateTime.Now` and `DateTimeOffset.Now`.
2. [Personally preferred, but breaking] Leave it alone. Maybe add a specific error message on `DateTimeOffset.Add(TimeSpan)` and `DateTime.Add(TimeSpan)` to make it more obvious. Pros: consistency. Major con: breaking change.

I want to choose 2, but I think the right answer is 1. Thoughts?

**Comment by MaceWindu** [comment]
I would expect it to work like this:
- if expression could be translated to sql - generate sql
- otherwise if it could be evaluated on client - create parameter
- if none above - throw `cannot be converted to SQL` error

Actually I would prefer to have it as generic translation logic, not tied to this specific case. @sdanyliv do we have anything like that inplace already?

Quick sketch design: for `Call` translation (`DTO.Add` in out case) if call supported for both client and server we check if we support server translation for all parameters including `this`, then if failed, check client transaltion support, otherwise throw exception. Does it make sense to implement like that?

**Comment by viceroypenguin** [comment]
Different question: As a user, should it be obvious that `DateTimeOffset.Now.AddMinute(1)` runs on the server, while `DateTimeOffset.Now.Add(new TimeSpan(hour: 0, minute: 1, second: 0)` runs on the client?

**Comment by MaceWindu** [comment]
When user really care about it - he use Sql.AsSql for server-side and variable for client-side evaluation :)

**Comment by viceroypenguin** [comment]
Ha. I've never used Sql.AsSql myself, so there's that.... :)

**Comment by viceroypenguin** [comment]
I'm good with your proposal. I'll start looking into it. It'll be a separate fix from #2514 then

**Comment by viceroypenguin** [comment]
@MaceWindu NB: That is how it works for Select queries. The current (3.1.3) code downloads the sysdatetimeoffset() from the Sql server and then calls `.Add()` on the client side. 

Inserts/Updates are different due to needing the full value before sending to the server. I'll revert the DTO.Now change for now to fix for 3.1.4 and continue proper research in 3.2.0. I'm not yet sure how to manage that difference.
